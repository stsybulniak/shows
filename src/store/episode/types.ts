export enum EpisodeTypes {
  fetchEpisodeList = '@@episode/FETCH_EPISODE_LIST',
  fetchEpisodeListSuccess = '@@episode/FETCH_EPISODE_LIST_SUCCESS',
  fetchEpisodeListFailure = '@@episode/FETCH_EPISODE_LIST_FAILURE',
}

export interface IEpisode {
  id: number;
  name: string;
  image: { medium: string; original: string } | null;
  summary: string;
  number: number;
  season: number;
}

export interface IEpisodeState {
  episodes: { [key: number]: IEpisode };
  isLoading: boolean;
  error: string | null;
}
