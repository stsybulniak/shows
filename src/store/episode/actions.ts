import { action } from 'typesafe-actions';
import { EpisodeTypes, IEpisode } from './types';

export const fetchEpisodeList = () => action(EpisodeTypes.fetchEpisodeList);
export const fetchEpisodeListSuccess = (payload: { [key: string]: IEpisode }) =>
  action(EpisodeTypes.fetchEpisodeListSuccess, payload);
export const fetchEpisodeListFailure = (payload: string) => action(EpisodeTypes.fetchEpisodeListFailure, payload);
