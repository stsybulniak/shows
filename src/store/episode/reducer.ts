import { Reducer } from 'redux';
import { IEpisodeState, EpisodeTypes } from './types';

export const initialState: IEpisodeState = {
  isLoading: false,
  episodes: {},
  error: null,
};

export const episodeReducer: Reducer<IEpisodeState> = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case EpisodeTypes.fetchEpisodeList: {
      return { ...state, isLoading: true };
    }
    case EpisodeTypes.fetchEpisodeListSuccess: {
      return { ...state, isLoading: false, episodes: payload };
    }
    case EpisodeTypes.fetchEpisodeListFailure: {
      return { ...state, isLoading: false, episodes: {}, error: payload };
    }
    default: {
      return state;
    }
  }
};
