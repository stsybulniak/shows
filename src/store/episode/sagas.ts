import { all, call, fork, put, takeEvery } from 'redux-saga/effects';
import { IEpisode, EpisodeTypes } from './types';
import axios, { AxiosResponse } from 'axios';
import { fetchEpisodeListFailure, fetchEpisodeListSuccess } from './actions';
import { convertEpisodeToMap } from '../../utils';

const fetchEpisodes = () => axios.get(`${process.env.REACT_APP_API_URL}/episodes`)

export function* handleFetchEpisodeList() {
  try {
    const resp: AxiosResponse<IEpisode[]> = yield call(fetchEpisodes);
    const data = convertEpisodeToMap(resp.data);
    yield put(fetchEpisodeListSuccess(data));
  } catch (err) {
    yield put(fetchEpisodeListFailure(err.message));
  }
}

function* watchEpisodeSaga() {
  yield takeEvery(EpisodeTypes.fetchEpisodeList, handleFetchEpisodeList);
}

function* episodeSaga() {
  yield all([fork(watchEpisodeSaga)]);
}

export default episodeSaga;
