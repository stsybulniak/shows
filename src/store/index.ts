import { combineReducers } from 'redux';
import { all, fork } from 'redux-saga/effects';
import { connectRouter, RouterState } from 'connected-react-router';
import { History } from 'history';
import showSaga from './show/sagas';
import { IShowState, showReducer } from './show';
import episodeSaga from './episode/sagas';
import { episodeReducer, IEpisodeState } from './episode';

export interface IApplicationState {
  router: RouterState;
  show: IShowState;
  episode: IEpisodeState;
}

export const createRootReducer = (history: History) =>
  combineReducers({
    router: connectRouter(history),
    show: showReducer,
    episode: episodeReducer,
  });

export function* rootSaga() {
  yield all([fork(showSaga), fork(episodeSaga)]);
}
