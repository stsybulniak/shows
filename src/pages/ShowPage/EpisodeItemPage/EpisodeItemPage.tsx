import { FC, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import Episode from '../../../components/Episode/Episode';
import ErrorOrLoading from '../../../components/ErrorOrLoading/ErrorOrLoading';
import { IApplicationState } from '../../../store';
import * as episodeActions from '../../../store/episode/actions';
import { viewportWidth } from '../../../utils';

const EpisodeItemPage: FC = () => {
  const { id } = useParams<{ id: string }>();
  const { episodes, isLoading, error } = useSelector((state: IApplicationState) => state.episode);
  const episode = episodes[parseInt(id)];
  const dispatch = useDispatch();

  useEffect(() => {
    if (!Object.keys(episodes).length && !error) {
      dispatch(episodeActions.fetchEpisodeList());
    }
  }, [episodes, error, dispatch]);

  return (
    <ErrorOrLoading isLoading={isLoading} error={error}>
      {episode && <Episode episode={episode} viewportWidth={viewportWidth} />}
    </ErrorOrLoading>
  );
};

export default EpisodeItemPage;
