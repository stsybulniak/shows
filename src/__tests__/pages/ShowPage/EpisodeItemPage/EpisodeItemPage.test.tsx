import { render } from '@testing-library/react';
import EpisodeItemPage from '../../../../pages/ShowPage/EpisodeItemPage/EpisodeItemPage';
import * as reactRedux from 'react-redux';
import { EpisodeTypes } from '../../../../store/episode';

const episodeMock = {
  id: 1156862,
  name: 'Mini Golf Madness',
  season: 2,
  number: 8,
  image: {
    medium: 'https://static.tvmaze.com/uploads/images/medium_landscape/115/288994.jpg',
    original: 'https://static.tvmaze.com/uploads/images/original_untouched/115/288994.jpg',
  },
  summary: '<p>Buttercup takes her game.</p>',
};

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    id: '1156862',
  }),
}));

describe('<EpisodeItemPage />', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });

  it('Render initial episode item page', () => {
    const dispatchMock = jest.fn();
    useSelectorMock.mockReturnValue({ episodes: {}, isLoading: true, error: null });
    useDispatchMock.mockReturnValue(dispatchMock);
    render(<EpisodeItemPage />);
    expect(dispatchMock).toHaveBeenCalledWith(expect.objectContaining({ type: EpisodeTypes.fetchEpisodeList }));
  });

  it('Render episode item page with sucsessfull episodes fetch', () => {
    const dispatchMock = jest.fn();
    useSelectorMock.mockReturnValue({ episodes: { 1156862: episodeMock }, isLoading: false, error: null });
    useDispatchMock.mockReturnValue(dispatchMock);
    const { baseElement } = render(<EpisodeItemPage />);
    expect(baseElement).toMatchSnapshot();
  });

  it('Render episode item page with erro episodes fetch', () => {
    const dispatchMock = jest.fn();
    useSelectorMock.mockReturnValue({ episodes: {}, isLoading: false, error: 'some error' });
    useDispatchMock.mockReturnValue(dispatchMock);
    const { baseElement } = render(<EpisodeItemPage />);
    expect(baseElement).toMatchSnapshot();
  });
});
