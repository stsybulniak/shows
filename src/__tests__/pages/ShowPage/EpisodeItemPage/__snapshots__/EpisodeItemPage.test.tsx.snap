// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EpisodeItemPage /> Render episode item page with erro episodes fetch 1`] = `
<body>
  <div>
    <h3>
      some error
    </h3>
  </div>
</body>
`;

exports[`<EpisodeItemPage /> Render episode item page with sucsessfull episodes fetch 1`] = `
<body>
  <div>
    <div
      class="Episode"
    >
      <button
        class="Episode__backButton"
      >
        Back to list
      </button>
      <h1>
        Mini Golf Madness
      </h1>
      <img
        alt="Mini Golf Madness"
        src="https://static.tvmaze.com/uploads/images/original_untouched/115/288994.jpg"
      />
      <div
        class="Episode__description"
      >
        <p>
          Buttercup takes her game.
        </p>
      </div>
    </div>
  </div>
</body>
`;
